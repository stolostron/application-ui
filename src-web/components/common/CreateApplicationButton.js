/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.

 *******************************************************************************/
// Copyright (c) 2020 Red Hat, Inc.
'use strict'

import React, { Component } from 'react'
import msgs from '../../../nls/platform.properties'
import { AcmButton } from '@open-cluster-management/ui-components'
import { Tooltip } from '@patternfly/react-core'
import { Link } from 'react-router-dom'
import config from '../../../lib/shared/config'
import { canCreateActionAllNamespaces } from '../../../lib/client/access-helper'
import _ from 'lodash'

const path = `${config.contextPath}/create`

class CreateApplicationButton extends Component {
  constructor(props) {
    super(props)
    this.state = {
      canDisable: true
    }
  }

  componentDidMount() {
    canCreateActionAllNamespaces('applications', 'create', 'app.k8s.io').then(
      response => {
        const disabled = _.get(response, 'data.userAccessAnyNamespaces')
        this.setState({ canDisable: !disabled })
      }
    )
  }

  disableClick(e) {
    e.preventDefault()
  }

  render() {
    const { locale } = this.context
    const { canDisable } = this.state
    if (canDisable) {
      return (
        <Tooltip
          content={msgs.get('actions.create.application.access.denied', locale)}
          isContentLeftAligned
          position="bottom"
        >
          {this.renderButton(locale, canDisable)}
        </Tooltip>
      )
    } else {
      return this.renderButton(locale, canDisable)
    }
  }

  renderButton = (locale, canDisable) => {
    return (
      <Link
        to={{
          pathname: path,
          state: { cancelBack: true }
        }}
        key="create-application"
        onClick={canDisable ? this.disableClick : undefined}
      >
        <AcmButton
          id="CreateAppButton"
          variant="primary"
          isSmall
          isDisabled={canDisable}
          data-test-create-application={!canDisable}
        >
          {msgs.get('actions.create.application', locale)}
        </AcmButton>
      </Link>
    )
  };
}

export default CreateApplicationButton
