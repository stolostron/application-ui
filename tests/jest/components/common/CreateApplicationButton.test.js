// Copyright (c) 2020 Red Hat, Inc.
"use strict";

let mockUserAccessAnyNamespaces = true;

jest.mock("../../../../lib/client/apollo-client", () => ({
  getUserAccessAllNamespaces: jest.fn(variables => {
    return Promise.resolve({
      data: {
        userAccessAnyNamespaces: mockUserAccessAnyNamespaces
      },
      loading: false,
      networkStatus: 7,
      stale: false
    });
  })
}));

import React from "react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import CreateApplicationButton from "../../../../src-web/components/common/CreateApplicationButton";

describe("CreateApplicationButton", () => {
  mockUserAccessAnyNamespaces = true;
  const componentEnabled = renderer.create(
    <BrowserRouter>
      <CreateApplicationButton />
    </BrowserRouter>
  );

  it("renders correctly when enabled", () => {
    const buttonClasses = componentEnabled.root.findByType("button").props
      .className;
    expect(buttonClasses).not.toContain("pf-m-disabled");
    expect(buttonClasses).not.toContain("pf-m-aria-disabled");
  });

  mockUserAccessAnyNamespaces = false;
  const componentDisabled = renderer.create(
    <BrowserRouter>
      <CreateApplicationButton />
    </BrowserRouter>
  );

  it("renders correctly when disabled", () => {
    const buttonClasses = componentDisabled.root.findByType("button").props
      .className;
    expect(buttonClasses).toContain("pf-m-aria-disabled");
  });
});
