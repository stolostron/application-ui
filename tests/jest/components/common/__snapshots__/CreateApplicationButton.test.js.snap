// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateApplicationButton renders correctly when disabled 1`] = `
<a
  href="undefined/create"
>
  <button
    aria-disabled="true"
    class="pf-c-button pf-m-primary pf-m-aria-disabled pf-m-small"
    data-ouia-component-id="OUIA-Generated-Button-primary-2"
    data-ouia-component-type="PF4/Button"
    data-ouia-safe="true"
    data-test-create-application="false"
    id="CreateAppButton"
    type="button"
  >
    Create application
  </button>
</a>
`;

exports[`CreateApplicationButton renders correctly when enabled 1`] = `
<a
  href="undefined/create"
>
  <button
    aria-disabled="false"
    class="pf-c-button pf-m-primary pf-m-small"
    data-ouia-component-id="OUIA-Generated-Button-primary-3"
    data-ouia-component-type="PF4/Button"
    data-ouia-safe="true"
    data-test-create-application="true"
    id="CreateAppButton"
    type="button"
  >
    Create application
  </button>
</a>
`;
